pkg/elle/core/graph.go

```diff
@@ -19,75 +19,263 @@ type DirectedGraph struct {
 
 // Vertices returns the set of all vertices in graph
 func (g *DirectedGraph) Vertices() []Vertex {
-	panic("impl me")
+	vertices := []Vertex{}
+
+	for key := range g.Outs {
+		vertices = append(vertices, key)
+	}
+	return vertices
 }
 
-// In returns inbound edges to v in graph g
+// In returns inbound vertices to v in graph g
 func (g *DirectedGraph) In(v Vertex) []Vertex {
-	panic("impl me")
+	_, ok := g.Ins[v]
+	if !ok {
+		return []Vertex{}
+	} else {
+		return g.Ins[v]
+	}
 }
 
-// Out returns outbound edges from v in graph g
+// Out returns outbound vertices from v in graph g
 func (g *DirectedGraph) Out(v Vertex) []Vertex {
-	panic("impl me")
+	_, ok := g.Outs[v]
+	if !ok {
+		return []Vertex{}
+	}
+
+	vertices := []Vertex{}
+
+	for key := range g.Outs[v] {
+		vertices = append(vertices, key)
+	}
+	return vertices
 }
 
 // Edges returns the edge between two vertices
 func (g *DirectedGraph) Edges(a, b Vertex) []Edge {
-	panic("impl me")
+	_, oka := g.Outs[a]
+	_, okb := g.Ins[b]
+	if !oka || !okb {
+		return []Edge{}
+	}
+
+	_, ok := g.Outs[a][b]
+	if !ok {
+		return []Edge{}
+	}
+
+	edges := []Edge{}
+	for _, rel := range g.Outs[a][b] {
+		edges = append(edges, Edge{
+			From:  a,
+			To:    b,
+			Value: rel,
+		})
+	}
+	return edges
 }
 
 // Link links two vertices relationship
-func (g *DirectedGraph) Link(v Vertex, succ []Vertex, rel Rel) {
-	panic("impl me")
+func (g *DirectedGraph) Link(v Vertex, succ Vertex, rel Rel) {
+	_, ok := g.Outs[v]
+	if !ok {
+		g.Outs[v] = make(map[Vertex][]Rel)
+	}
+	_, ok = g.Outs[succ]
+	if !ok {
+		g.Outs[succ] = make(map[Vertex][]Rel)
+	}
+
+	_, ok = g.Ins[v]
+	if !ok {
+		g.Ins[v] = []Vertex{}
+	}
+	_, ok = g.Ins[succ]
+	if !ok {
+		g.Ins[succ] = []Vertex{}
+	}
+
```

Comment:
https://paste.ubuntu.com/p/fM69YNN6sq/

Actually I don't think you need to initialize it.

Reply:
No, we need ensure all keys at `Outs` are also at `Ints` and vice versa. @mapleFU

Reply:
Does this function add an edge from `v` to `succ`? If so, what's the purpose of creating `g.Outs[succ]` and `g.Ins[v]`?

Reply:
If there is no initialization, the points without in-degrees are not in "Ins". Similarly, the points without out-degrees are not in "Outs". In this case, we need to modify "Vertices()".